wasm.js

```diff
@@ -0,0 +1,178 @@
+ *
+ * 0chain @zerochain/0chain program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+'use strict';
+
+const utils = require('./utils');
+const g = global || window || self;
+
+async function blsSign(hash) {
+    if (!bridge.js?.secretKey) {
+        const errMsg = 'err: bls.secretKey is not initialized';
+        console.error(errMsg);
+        throw new Error(errMsg);
+    }
+
+    const bytes = utils.hexStringToByte(hash);
+
+    const sig = bridge.js.secretKey.sign(bytes);
+
+    if (!sig) {
+        const errMsg = 'err: wasm blsSign function failed to sign transaction';
+        console.error(errMsg);
+        throw new Error(errMsg);
+    }
+
+    return sig.serializeToHexStr();
+}
+
+
+// Initialize __zcn_wasm__ 
+if (!g.__zcn_wasm__) {
+    g.__zcn_wasm__ = {
+        js: {
+            secretKey: null,
+            sign: blsSign,
+        },
+        sdk: {},
+    };
+}
+
+/**
+ * bridge is an easier way to refer to the Go WASM object.
+ */
+const bridge = g.__zcn_wasm__;
+
+/**
+ * Sleep is used when awaiting for Go Wasm to initialize.
+ * It uses the lowest possible sane delay time (via requestAnimationFrame).
+ * However, if the window is not focused, requestAnimationFrame never returns.
+ * A timeout will ensure to be called after 50 ms, regardless of whether or not the tab is in focus.
+ *
+ * @returns {Promise} an always-resolving promise when a tick has been completed.
+ */
+const sleep = () => new Promise(res => {
+    requestAnimationFrame(res);
+    setTimeout(res, 1000);
+});
+
+/**
+ * The maximum amount of time that we would expect Wasm to take to initialize.
+ * If it doesn't initialize after this time, we send a warning to console.
+ * Most likely something has gone wrong if it takes more than 3 seconds to initialize.
+ */
+const maxTime = 10 * 1000;
+
+async function setWallet(bls, clientID, sk, pk) {
+    bridge.js.secretKey = bls.deserializeHexStrToSecretKey(sk);
+    //use proxy.sdk to detect if sdk is ready
+    await bridge.__proxy__.sdk.setWallet(clientID, pk)
+    // await bridge.sdk.setWallet(clientID, pk)
+}
+
+const jsProxy = new Proxy(
+    {},
+    {
+        get: (_, key) => bridge.js[key],
+        set: (_, key, value) => {
+            bridge.js[key] = value;
+        },
+    }
+)
+
+async function loadWasm(go) {
+    const result = await WebAssembly.instantiateStreaming(
+        await fetch('/zcn.wasm'),
+        go.importObject
+    );
+    go.run(result.instance);
```

Comment:
This way you can discount the time it took for the wasm binary to be fetched, which can vary greatly.

```suggestion
    const result = await WebAssembly.instantiateStreaming(
        await fetch('/zcn.wasm'),
        go.importObject
    );
    
    setTimeout(() => {
        if (bridge.__wasm_initialized__ !== true) {
            console.warn(
                'wasm bridge (zcn.__wasm_initialized__) still not true after max time'
            );
        }
    }, maxTime);
    
    go.run(result.instance);
```